xfrd-catalog-zones.c

```diff
@@ -0,0 +1,1275 @@
+	                    , ((struct catalog_member_zone*)right)->member_id);
+}
+
+static struct xfrd_catalog_producer_zone*
+xfrd_get_catalog_producer_zone(struct catalog_member_zone* cmz)
+{
+	struct zone_options *producer_zopt;
+	struct xfrd_catalog_producer_zone* producer_zone;
+	const dname_type* producer_name;
+	const char* producer_name_str;
+
+	assert(xfrd);
+	if(!cmz || !cmz->options.pattern->catalog_producer_zone)
+		return NULL;
+
+	/* TODO: Store as dname in pattern->catalog_producer_zone */
+	producer_name = dname_parse(xfrd->nsd->options->region,
+			cmz->options.pattern->catalog_producer_zone);
+	producer_zopt = zone_options_find(xfrd->nsd->options, producer_name);
+	producer_name_str = dname_to_string(producer_name, NULL);
+	region_recycle( xfrd->nsd->options->region, (void *)producer_name
+	              , dname_total_size(producer_name));
+	if(!producer_zopt) {
+		log_msg(LOG_ERR, "catalog producer zone '%s' not found for "
+			"zone '%s'", producer_name_str, cmz->options.name);
+		return NULL;
+	}
+	if(!zone_is_catalog_producer(producer_zopt)) {
+		log_msg(LOG_ERR, "cannot add catalog producer member "
+			"zone '%s' to non producer zone '%s'",
+			cmz->options.name, producer_zopt->name);
+		return NULL;
+	}
+	producer_name = (dname_type*)producer_zopt->node.key;
+	producer_zone = (struct xfrd_catalog_producer_zone*)
+		rbtree_search(xfrd->catalog_producer_zones, producer_name);
+	if (!producer_zone) {
+		/* Create a new one */
+		DEBUG(DEBUG_XFRD, 1, (LOG_INFO,"creating catalog producer zone"
+			" '%s'", producer_zopt->name));
+		producer_zone = (struct xfrd_catalog_producer_zone*)
+			region_alloc(xfrd->region, sizeof(*producer_zone));
+		memset(producer_zone , 0, sizeof(*producer_zone));
+		producer_zone->node.key = producer_zopt->node.key;
+		producer_zone->options = producer_zopt;
+		producer_zone->member_ids.region = xfrd->region;
+		producer_zone->member_ids.root = RBTREE_NULL;
+		producer_zone->member_ids.count = 0;
+		producer_zone->member_ids.cmp = member_id_compare;
+		producer_zone->serial = 0;
+		producer_zone->to_delete = NULL;
+		producer_zone->to_add = NULL;
+		producer_zone->latest_pxfr = NULL;
+		producer_zone->axfr = 1;
+		rbtree_insert(xfrd->catalog_producer_zones,
+				(rbnode_type*)producer_zone);
+	}
+	return producer_zone;
+}
+
+void
+xfrd_add_catalog_producer_member(struct catalog_member_zone* cmz)
+{
+	struct xfrd_catalog_producer_zone* producer_zone;
+	const dname_type* producer_name;
+	struct xfrd_producer_member* to_add;
+
+	assert(xfrd);
+	if (!(producer_zone = xfrd_get_catalog_producer_zone(cmz))) {
+		return;
+	}
+	producer_name = producer_zone->node.key;
+	while(!cmz->member_id) {
+		/* Make new member_id with this catalog producer */
+		char id_label[sizeof(uint32_t)*2+1];
+		uint32_t new_id = (uint32_t)random_generate(0x7fffffff);
+
+		id_label[hex_ntop( (void *)&new_id
+		                 , sizeof(uint32_t)
+		                 , id_label
+		                 , sizeof(id_label))] = 0;
```

Comment:
Please don't do this. We expect a fixed constant to be returned. `2 * srclength` to be precise. Please just use that when null-terminating.

Reply:
resolved in 5a4b474a